weave/trace/weave_client_send_file_cache.py

```diff
@@ -0,0 +1,171 @@
+"""
+In memory cache of files sent to the server to avoid sending the same file
+multiple times, using an LRU (Least Recently Used) eviction policy.
+"""
+
+import hashlib
+import threading
+from collections import OrderedDict
+from concurrent.futures import Future
+from typing import Generic, Optional, TypeVar
+
+from weave.trace_server.trace_server_interface import FileCreateReq, FileCreateRes
+
+# Define generic type variables
+K = TypeVar("K")  # Key type
+V = TypeVar("V")  # Value type
+
+
+class ThreadSafeLRUCache(Generic[K, V]):
```

Comment:
Isn't `functools.lru_cache` already thread safe?  Couldn't you just decorate:
```py
@lru_cache(...)
def _send_file_create(...)
```

Reply:
https://docs.python.org/3/library/functools.html#functools.lru_cache

Reply:
In general I prefer the builtins because they are battle-tested and should be faster. 
 Did you consider wrapping the builtin `lru_cache`?

Otherwise, I'll add comments that assume you go with this impl

Reply:
lru_cache is insufficient here since in a multi-threaded context, the same file can be uploaded multiple times

Reply:
But can you wrap `lru_cache` with locks/conditions?